kartothek/io_components/read.py

```diff
@@ -133,35 +124,49 @@ def dispatch_metapartitions_from_factory(
             )
 
 
-def _allowed_labels_by_predicates(predicates, dataset_factory, dispatch_by):
-    if len(predicates) == 0:
-        raise ValueError("The behaviour on an empty list of predicates is undefined")
-
-    dataset_factory = dataset_factory.load_partition_indices()
-
+def _columns_in_predicates(predicates):
     # Determine the set of columns that are part of a predicate
     columns = set()
     for predicates_inner in predicates:
         if len(predicates_inner) == 0:
             raise ValueError("The behaviour on an empty predicate is undefined")
         for col, _, _ in predicates_inner:
             columns.add(col)
+    return columns
+
+
+def _load_indices_dispatch_by(
+    dispatch_by: List[str], dataset_factory: DatasetFactory
+) -> None:
+    for col in dispatch_by:
+        if col in dataset_factory.partition_keys:
+            dataset_factory.load_partition_indices()
+        else:
+            dataset_factory.load_index(col)
+
+
+def _load_indices_predicates(predicates, dataset_factory):
+    columns = _columns_in_predicates(predicates)
+
+    if set(dataset_factory.partition_keys).union(columns):
+        dataset_factory = dataset_factory.load_partition_indices()
 
     # Load the necessary indices
     for column in columns:
         if column in dataset_factory.indices:
             dataset_factory = dataset_factory.load_index(column)
+    return dataset_factory
```

Comment:
why does this function return the factoy and `_load_indices_dispatch_by` does not? Alos, all new functions here missing docstrings

Reply:
Refactored such that this function is not used anymore